### PR TITLE
arch/arm: Remove the code copy register from xcpt to stack

### PR DESCRIPTION
**Pull request by xiaoxiang781216**
## Summary
since xcpt always point to the stack after the below change(#5645):
commit 7b9978883c46569068ba5dac5755f998fbd237b2
Author: chao.an <anchao@xiaomi.com>
Date:   Tue Mar 1 01:06:24 2022 +0800

    arch/arm: optimize context switch speed

    The current context save implementation saves registers of each task
    to xcp context, which is unnecessary because most of the arm registers are
    already saved in the task stack, this commit replace the xcp context with
    stack context to improve context switching performance and reduce the tcb
    space occupation of tcb instance.

Signed-off-by: Xiang Xiao <xiaoxiang@xiaomi.com>

## Impact
Improve the context switch performance(reduce copy 8 integer + 16 float) on armv6-m, armv7-m and armv8-m.

### CODE REVIEW

**Review by pkarashchenko — APPROVED** [review]

**Comment by masayuki2009** [comment]
@xiaoxiang781216 

Hmm, ostest with lm3s6965-ek:qemu-protected crashed.

```
$ ~/opensource/QEMU/qemu-5.2/build/arm-softmmu/qemu-system-arm -M lm3s6965evb -device loader,file=./nuttx.bin,addr=0x00000000 -device loader,file=./nuttx_user.bin,addr=0x00020000 -nographic
ABCDEF
stdio_test: write fd=1
stdio_test: Standard I/O Check: printf
stdio_test: write fd=2
stdio_test: Standard I/O Check: fprintf to stderr
ostest_main: Started user_main at PID=2

...

End of test memory usage:
VARIABLE  BEFORE   AFTER
======== ======== ========
arena        7eb0     7eb0
ordblks         1        1
mxordblk     5b00     52e0
uordblks     23b0     2bd0
fordblks     5b00     52e0

user_main: vfork() test
up_assert: Assertion failed at file:armv7-m/arm_hardfault.c line: 173 task: ostest
arm_registerdump: R0: 00000000 R1: 00000000 R2: 00000000  R3: deadbeef
arm_registerdump: R4: 2000945c R5: 20004018 R6: 00026d05  FP: 00000000
arm_registerdump: R8: 20008cb8 SB: 0002686c SL: 00000000 R11: 00000000
arm_registerdump: IP: 00000000 SP: 20009e68 LR: 00000000  PC: 00006190
arm_registerdump: xPSR: 00000200 PRIMASK: 00000000 CONTROL: 00000000
arm_registerdump: EXC_RETURN: fffffff9
arm_dump_stack: User Stack:
arm_dump_stack: sp:     20009d78
arm_dump_stack:   base: 200096a0
arm_dump_stack:   size: 000007c8
arm_stackdump: 20009d60: 00000000 200096a0 20009e68 20009d78 20000268 000046eb 00000000 00000000
arm_stackdump: 20009d80: 00006190 00000000 00006190 0000c5d6 00000003 00000000 20008cb8 00003927
arm_stackdump: 20009da0: 00000000 00000ea3 00000000 20000268 00000003 00001313 00000000 00000b65
arm_stackdump: 20009dc0: 2000945c 20004018 00026d05 000004f7 00006b40 fffffffd 00020135 00000005
arm_stackdump: 20009de0: 20008cb8 00000000 00000000 00000000 0002004d 20008cb8 00001c11 00000000
arm_stackdump: 20009e00: 00001c11 00000000 00000000 00000000 00000000 deadbeef 20009e68 00000000
arm_stackdump: 20009e20: 2000945c 20004018 00026d05 00000000 20008cb8 0002686c 00000000 00000000
arm_stackdump: 20009e40: fffffff9 00000000 00000000 00000000 deadbeef 00000000 00000000 00006190
arm_showtasks:    PID    PRI      USED     STACK   FILLED    COMMAND
arm_dump_task:      0      0       328      1000    32.8%    Idle Task
arm_dump_task:      1    100       600      2024    29.6%    ostest_main
arm_dump_task:      2    100       912      1992    45.7%    ostest
arm_dump_task:     68    100       560      1992    28.1%    ostest

```

**Comment by xiaoxiang781216** [comment]
@masayuki2009 https://github.com/apache/incubator-nuttx/pull/5774 forget to update vfork'sp, @anchao will provide a patch soon.

**Comment by anchao** [comment]
> Hmm, ostest with lm3s6965-ek:qemu-protected crashed.

@masayuki2009 , please rebase this PR again

File: arch/arm/src/common/arm_vfork.c [hunk 1]
```diff
@@ -137,10 +137,8 @@ pid_t up_vfork(const struct vfork_s *context)
    * effort is overkill.
    */
 
-  newtop = STACK_ALIGN_DOWN((uint32_t)child->cmn.stack_base_ptr +
-                                      child->cmn.adj_stack_size -
-                                      XCPTCONTEXT_SIZE);
-  newsp = newtop - stackutil;
+  newsp = (uint32_t)child->cmn.stack_base_ptr + child->cmn.adj_stack_size;
+  newtop = (uint32_t)newsp - XCPTCONTEXT_SIZE;
```
Comment (xiaoxiang781216) [inline arch/arm/src/common/arm_vfork.c:141]: remove (uint32_t)

File: arch/arm/src/common/arm_vfork.c [hunk 2]
```diff
@@ -137,10 +137,19 @@ pid_t up_vfork(const struct vfork_s *context)
...
+  newsp = (uint32_t)child->cmn.stack_base_ptr +
+                    child->cmn.adj_stack_size - stackutil;
+
+  newtop = newsp - XCPTCONTEXT_SIZE;
+
+  /* Move the register context to newtop. */
+
+  memcpy((void *)newtop, child->cmn.stack_base_ptr +
+                         child->cmn.adj_stack_size -
+                         XCPTCONTEXT_SIZE, XCPTCONTEXT_SIZE);
```
Comment (xiaoxiang781216) [inline arch/arm/src/common/arm_vfork.c:149]: "child->cmn.stack_base_ptr + child->cmn.adj_stack_size - XCPTCONTEXT_SIZE" to tcb->xcpt->regs

File: arch/arm/src/common/arm_vfork.c [hunk 3]
```diff
@@ -137,10 +137,19 @@ pid_t up_vfork(const struct vfork_s *context)
...
    */
 
-  newtop = STACK_ALIGN_DOWN((uint32_t)child->cmn.stack_base_ptr +
-                                      child->cmn.adj_stack_size -
-                                      XCPTCONTEXT_SIZE);
-  newsp = newtop - stackutil;
+  newsp = (uint32_t)child->cmn.stack_base_ptr +
+                    child->cmn.adj_stack_size - stackutil;
+
+  newtop = newsp - XCPTCONTEXT_SIZE;
```
Comment (xiaoxiang781216) [inline arch/arm/src/common/arm_vfork.c:143]: newtop->newregs

File: arch/arm/src/common/arm_vfork.c (hunk 2 above)
Comment (anchao) [reply arch/arm/src/common/arm_vfork.c]: Done

File: arch/arm/src/common/arm_vfork.c (hunk 3 above)
Comment (anchao) [reply arch/arm/src/common/arm_vfork.c]: Done

**Comment by masayuki2009** [comment]
@anchao @xiaoxiang781216 

 lm3s6965-ek:qemu-protected works now.

However, I noticed that a deadlock happens with raspberrypi-pico:smp.
Without this PR, the issue does not happen.

```
NuttShell (NSH) NuttX-3.6.1
nsh> uname -a
NuttX 3.6.1 ae6658ba6e Mar 18 2022 23:27:18 arm raspberrypi-pico
nsh> ps
  PID GROUP CPU PRI POLICY   TYPE    NPX STATE    EVENT     SIGMASK   STACK   USED  FILLED COMMAND
    0     0   0   0 FIFO     Kthread N-- Assigned           00000000 001000 000416  41.6%  CPU0 IDLE
    1     1   1   0 FIFO     Kthread N-- Running            00000000 001000 000112  11.2%  CPU1 IDLE
    2     2   0 100 RR       Task    --- Running            00000000 002024 001168  57.7%  nsh_main
nsh> free
                   total       used       free    largest  nused  nfree
        Umem:     257728       9408     248320     248320     37      1
nsh> ostest
stdio_test: write fd=1
stdio_test: Standard I/O Check: printf
stdio_test: write fd=2
...
End of test memory usage:
VARIABLE  BEFORE   AFTER
======== ======== ========
arena       3eec0    3eec0
ordblks         2        4
mxordblk    38490    37bc0
uordblks     69e0     69e0
fordblks    384e0    384e0

user_main: pthread_rwlock test
pthread_rwlock: Initializing rwlock
```

**Comment by masayuki2009** [comment]
@anchao 

>The same issue also happens with raspberrypi-pico:nsh (non SMP).

It seems that the hardfault happened.

```
End of test memory usage:
VARIABLE  BEFORE   AFTER
======== ======== ========
arena       407d0    407d0
ordblks         4        5
mxordblk    39ba0    39ba0
uordblks     5180     5180
fordblks    3b650    3b650

user_main: pthread_rwlock test
pthread_rwlock: Initializing rwlock

b up_assert                                                                                                                                                  
Breakpoint 1 at 0x10004b60: file common/arm_assert.c, line 479.                                                                                              
Note: automatically using hardware breakpoints for read-only addresses.                                                                                      
(gdb) c                                                                                                                                                      
Continuing.                                                                                                                                                  
                                                                                                                                                             
Thread 1 hit Breakpoint 1, up_assert (filename=filename@entry=0x10015780 "armv6-m/arm_hardfault.c", lineno=lineno@entry=123) at common/arm_assert.c:479      
479       syslog_flush();                                                                                                                                    
(gdb) where                                                                                                                                                  
#0  up_assert (filename=filename@entry=0x10015780 "armv6-m/arm_hardfault.c", lineno=lineno@entry=123) at common/arm_assert.c:479                             
#1  0x10003772 in _assert (filename=filename@entry=0x10015780 "armv6-m/arm_hardfault.c", linenum=linenum@entry=123) at assert/lib_assert.c:36                
#2  0x10000a5a in arm_hardfault (irq=<optimized out>, context=<optimized out>, arg=<optimized out>) at armv6-m/arm_hardfault.c:123                           
#3  0x10001496 in irq_dispatch (irq=irq@entry=3, context=<optimized out>) at irq/irq_dispatch.c:179                                                          
#4  0x100008ec in arm_doirq (irq=3, regs=0x0) at armv6-m/arm_doirq.c:70                                                                                      
#5  0x10000514 in exception_common () at armv6-m/arm_exception.S:188                                                                                         
Backtrace stopped: previous frame identical to this frame (corrupt stack?)              
```

**Comment by xiaoxiang781216** [comment]
@masayuki2009 could you try the new version?

**Comment by anchao** [comment]
>>The same issue also happens with raspberrypi-pico:nsh (non SMP).
>It seems that the hardfault happened.

@masayuki2009 san,

I have fix this crash, please rebase this PR and try again!

The following 5 configs works fine with this PR:
sabre-6quad/netknsh (/mnt/nfs/bin/getprime)
sabre-6quad:smp (ostest)
lm3s6965-ek:qemu-kostest (ostest)
lm3s6965-ek:qemu-protected (ostest)
raspberrypi-pico:nsh (ostest)

File: arch/arm/src/armv7-a/arm_schedulesigaction.c [hunk 4]
```diff
@@ -326,13 +329,14 @@ void up_schedule_sigaction(struct tcb_s *tcb, sig_deliver_t sigdeliver)
                    * been delivered.
                    */
 
-                  CURRENT_REGS            =
-                    (FAR void *)STACK_ALIGN_DOWN((uint32_t)CURRENT_REGS -
-                                                 (uint32_t)XCPTCONTEXT_SIZE);
+                  CURRENT_REGS            = (FAR void *)
```
Comment (no1wudi) [inline arch/arm/src/armv7-a/arm_schedulesigaction.c:332]: Remove all `FAR`?

Comment (xiaoxiang781216) [reply arch/arm/src/armv7-a/arm_schedulesigaction.c:332]: FAR is still used in arm, it's better to remove all FAR in one patch.

**Comment by masayuki2009** [comment]
>I have fix this crash, please rebase this PR and try again!

@anchao 
I confirmed that the latest PR fixed all of the issues.

**Pull request merged**